Specify the mapping for Prometheus and Statsd exporters (#118)
open-telemetry/oteps: text/metrics/0118-statsd-and-prometheus-exporters.md

these OSS exposition formats, in order to support migration from
Prometheus and Statsd APIs without changing metrics protocols.
Specifying the data type correspondence is a necessary prerequisite
for migration from Prometheus and Statsd instruments to OpenTelemetry
instruments.  The instrument migration is included to complete this
proposal.

For Counter and Gauge instruments, this proposal promises that the
default mapping from Prometheus and Statsd instrument to OpenTelemetry
instrument, paired with the default Aggregator and then mapped back
into Prometheus or Statsd, produces the corresponding Counter or Gauge
value type of the original Prometheus or Statsd instrument.  The same
is not true for Histogram and Summary instruments: the recommended
OpenTelemetry instruments will export Counter or Gauge values by
default.  It is expected that a configurable SDK or the Metric Views
API will be used to reconfigure selected OpenTelemetry instruments to
produce Histogram and Summary values in Prometheus and Statsd.

## Background

An Aggregator is an implementation of some logic to compute an
Aggregation, which is an exact or approximate summarization of a

> series of metric events.  Exporters translate Aggregation values to
> into an exposition format, so the choice of Aggregator decides which

Reviewer: typo?
```suggestion
series of metric events.  Exporters translate Aggregation values
into an exposition format, so the choice of Aggregator decides which
```

These properties will help understand how to map OpenTelemetry
Aggregators into Prometheus and Statsd metric data.  These are the
OpenTelemetry instruments:

| Name | Synchronous | Adding or Grouping | Monotonic | Precomputed-Sum | Default Aggregator |
| ---- | ----------- | -------- | --------- | ---- | --- |
| Counter           | Yes | Adding   | Yes | No  | Sum |
| UpDownCounter     | Yes | Adding   | No  | No  | Sum |
| ValueRecorder     | Yes | Grouping | n/a | n/a | MinMaxLastSumCount |
| SumObserver       | No  | Adding   | Yes | Yes | Sum |
| UpDownSumObserver | No  | Adding   | No  | Yes | Sum |
| ValueObserver     | No  | Grouping | n/a | n/a | MinMaxLastSumCount |

Note that the Precomputed-Sum property places some constraints on the
combination of Aggregator and Exporter.  To compute _Delta_
aggregations of Precomputed-Sum instruments requires an aggregation
that supports subtraction (which MUST include Sum and SHOULD include
Histogram).

### Prometheus

This is based off of the Prometheus [Metric
Types](https://prometheus.io/docs/concepts/metric_types).

#### Prometheus instruments

Prometheus Counter instruments are semantically identical to

> OpenTelemetry Counter instruments, including the Mnotonic property.

Reviewer 2: ```suggestion
OpenTelemetry Counter instruments, including the Monotonic property.
```

`Add()` map into OpenTelemetry UpDownCounter instruments (maybe
SumObserver, UpDownSumObserver), while Prometheus Gauges used to
`Set()` map into OpenTelemetry ValueRecorder instruments (maybe
ValueObserver).

Prometheus Histogram instruments are exposed as Cumulative
aggregations, defined by a number of bucketed counts, accumulated from
the start of the process.  Prometheus Histogram instruments map into
ValueRecorder instruments (maybe ValueObserver).  To configure a
Histogram exposition in the Prometheus exporter, configure a Histogram
Aggregator with the desired buckets.

Prometheus Summary instruments are discouraged, as mentioned above,
because they are not mergeable.  Uses of Prometheus Summary
instruments map into OpenTelemetry ValueRecorder instruments (maybe
ValueObserver).  To configure a Summary exposition in the Prometheus
exporter, configure a Sketch Aggregator with the desired quantiles.

#### OpenTelemetry Aggregator to Prometheus exposition

The following table lists the mapping from Aggregator to Prometheus
exposition format.  The "Typical Instruments" listed are the
applicable OpenTelemetry instruments, for which the Prometheus mapping
is sensible.

> | OpenTelemetry aggregator | Default Prometheus data type mapping | Export kind | Typical instruments | Notes |

Reviewer 2: I see potentially two different people reading this table: an OpenTelemetry developer writing an exporter for prometheus, and an existing Prometheus instrumentation author looking to migrate to OpenTelemetry instrumentation.

This table seems to serve the the former well, but I'm not sure it is clear for the latter as to what they need to do. Guessing there would like a more explicit mapping between Prometheus type and OTel Instrument. I'm guessing the one-to-many mapping might be a bit overwhelming.

Is this something we can assume Prometheus instrumentation writers can work out (i.e. become familiar with synchronicity, precomputed sums, and additive/grouping)?

Is this something for a different PR?

Could this be added in a subsequent table?

| -- | -- | -- | -- | -- |

> | Sum (Monotonic)       | Counter    | Cumulative | Counter(*), SumObserver(*) | |
> | Sum (Non-Monotonic)   | Gauge      | Cumulative | UpDownCounter(*), UpDownSumObserver(*) | |
> | LastValue             | Gauge      | Cumulative | ValueRecorder, ValueObserver, SumObserver, UpDownSumObserver | |
> | MinMaxLastSumCount    | Gauge      | Delta      | ValueRecorder(*), ValueObserver(*) | Expose the LastValue field as the Gauge |

Reviewer 2: ```suggestion
| Sum (Monotonic)       | Counter    | Cumulative | Counter(\*), SumObserver(\*) | |
| Sum (Non-Monotonic)   | Gauge      | Cumulative | UpDownCounter(\*), UpDownSumObserver(\*) | |
| LastValue             | Gauge      | Cumulative | ValueRecorder, ValueObserver, SumObserver, UpDownSumObserver | |
| MinMaxLastSumCount    | Gauge      | Delta      | ValueRecorder(\*), ValueObserver(\*) | Expose the LastValue field as the Gauge |
```

> | MinMaxLastSumCount    | Gauge      | Delta      | ValueRecorder(*), ValueObserver(*) | Expose the LastValue field as the Gauge |

Reviewer 3: The notes say that LastValue will be exposed as a Gauge.

Is the idea that we will throw away the other non-last value (min/max/sum/count) by default?

I would have thought a more natural default would be for each of these aggregations to be represented as a separate Gauge, otherwise we are aggregating that data just to throw it away - i.e.

```yaml
metricname_min
metricname_max
metricname_last
metricname_sum
metricname_count
```

That would obviously have some implications for the mapping from Prometheus -> OT -> Prometheus that I'm not too sure how to address.

Author: What you propose is a viable export strategy, I just don't believe it's the one most Prometheus users want or expect. This is what the Prometheus "Summary" metric exports, approximately. If we replaced "metricname_last" with just "metricname" that would create less confusion for Prometheus users, but they might not appreciate the cost of indexing new metric names that they never wanted.

It would be more optimal, but much more complex, to try to match the default aggregation to the configured exporter. If an OTLP exporter is configured, then the question becomes recursive. What aggregation should I use to satisfy a downstream exporter? That would create a headache at startup.

The idea of adding an optimization (#117) to store a single value when there is in fact a single value, was meant to address some of the concern about throwing away information. I believe some platforms want to see min/max/sum/count and would not like to force users of those systems to reconfigure the aggregation.

Reviewer 3: Makes sense. Thanks for the detailed explanation

| Histogram             | Histogram  | Cumulative | ValueRecorder, ValueObserver | |
| Sketch                | Summary    | Delta      | ValueRecorder, ValueObserver | |
| Exact                 | Summary    | Delta      | ValueRecorder, ValueObserver | |

> Above, (*) denotes the default behavior of an OpenTelemetry

Reviewer 2: ```suggestion
Above, (\*) denotes the default behavior of an OpenTelemetry
```

instrument.

### Statsd

Statsd refers to a wire format for metrics. The mapping specified here
refers to either the original Etsy protocol (a.k.a. plain Statsd) or
the DataDog variation with labels added (a.k.a. DogStatsd).

> #### Statsd instruments

Reviewer 4: @jmacd I believe this section is outlining the `statsdreceiver` metric type mappings? I see Summary in the metrics.proto, but I don't see Sketch. Can you share more details on the OpenTelemetry Sketch aggregator?

Author: There is no OTLP support for a sketch aggregator at this time. It is specified as an option for the SDK as a means of computing Summary values, although we would like to include OTLP support for sketches there is research and protocol description work to be done. Datadog has not published a serialization format for DDSketch as far as I know. I've proposed the use of a T-digest compatible format.

See this thread: https://github.com/open-telemetry/oteps/pull/113#discussion_r441885123

Statsd instruments export individual data points using messages
described by a 1- or 2- character string:

- "c" for Counter
- "g" for Gauge
- "h" for Histogram (exposed as a Summary)
- "ms" for Timing (exposed as a Summary)
- "d" for Distribution (exposed as a Sketch)

#### OpenTelemetry Aggregator to Statsd exposition

The Statsd exposition format always uses Delta aggregation.

> Statsd Grouping instruments, which are all except the Statsd Counter
> instrument, are exposed as Gauge by default (e.g., as opposed to

Reviewer 2: I'm not following the meaning here. Maybe?

```suggestion
Statsd Grouping instruments which are all, except the Statsd Counter
instrument, exposed as Gauges by default (e.g., as opposed to a
```

Histogram).

The Histogram, Sketch and Exact Aggregators, when configured, is
exposed in Summary form, using the instrument name with `_sum`,
`_count`, and various quantile suffixes (e.g., `_p95`).  The
MinMaxLastSumCount Aggregator also supports being exposed as a
Summary (e.g., `_min`, `_max` suffixes).

The following table lists the mapping from Aggregator to Statsd
exposition format.  The "Typical Instruments" listed are the
applicable OpenTelemetry instruments, for which the Statsd mapping is
sensible.

| OpenTelemetry aggregator | Default Statsd data type mapping | Typical instruments | Notes |
| -- | -- | -- | -- |

> | Sum (Monotonic)       | Counter    | Counter(*), SumObserver(*) | |
> | Sum (Non-Monotonic)   | Gauge      | UpDownCounter(*), UpDownSumObserver(*) | |
> | LastValue             | Gauge      | ValueRecorder, ValueObserver, SumObserver, UpDownSumObserver | |
> | MinMaxLastSumCount    | Gauge      | ValueRecorder(*), ValueObserver(*) | Expose the LastValue field as the Gauge |
> | Histogram             | Summary    | ValueRecorder, ValueObserver | |
> | Sketch                | Summary    | ValueRecorder, ValueObserver | |
> | Exact                 | Summary    | ValueRecorder, ValueObserver | |
> 
> Above, (*) denotes the default behavior of an OpenTelemetry

Reviewer 2: ```suggestion
| Sum (Monotonic)       | Counter    | Counter(\*), SumObserver(\*) | |
| Sum (Non-Monotonic)   | Gauge      | UpDownCounter(\*), UpDownSumObserver(\*) | |
| LastValue             | Gauge      | ValueRecorder, ValueObserver, SumObserver, UpDownSumObserver | |
| MinMaxLastSumCount    | Gauge      | ValueRecorder(\*), ValueObserver(\*) | Expose the LastValue field as the Gauge |
| Histogram             | Summary    | ValueRecorder, ValueObserver | |
| Sketch                | Summary    | ValueRecorder, ValueObserver | |
| Exact                 | Summary    | ValueRecorder, ValueObserver | |

Above, (\*) denotes the default behavior of an OpenTelemetry
```
